### PR TITLE
Add Pin/Unpin action in quick access of the message action bar

### PR DESCRIPTION
**Pull request by florianduros**
## Checklist

-   [X] Tests written for new code (and old code if feasible).
-   [X] New or updated `public`/`exported` symbols have accurate [TSDoc](https://tsdoc.org/) documentation.
-   [X] Linter and other CI checks pass.
-   [ ] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md)).

Related to https://github.com/element-hq/element-internal/issues/607.
Figma https://www.figma.com/design/nwbnk8RaZctjrrRGTUcx1X/Pinned-Messages?node-id=38-170276&t=QMMsCWBcCSoSiyn5-0

* Add pin/unpin action in quick access of the message action bar
* Use new pin/unpin icon
* Pin/unpin is available in the context menu of the message action bar only when right clicked (same behaviour than edit, threads etc)
* Move pinning function from context menu to `PinningUtils`

| Before  | After |
| ------------- | ------------- |
| Screenshot 2024-08-19 at 15 33 29| Screenshot 2024-08-19 at 15 31 20|
| Screenshot 2024-08-19 at 15 33 50| Screenshot 2024-08-19 at 15 31 32|
| Screenshot 2024-08-19 at 15 34 07|Screenshot 2024-08-19 at 15 31 58|

### CODE REVIEW

File: test/components/views/messages/MessageActionBar-test.tsx [hunk 1]
```diff
@@ -468,4 +471,33 @@ describe("<MessageActionBar />", () => {
...
+            jest.spyOn(
+                room.getLiveTimeline().getState(EventTimeline.FORWARDS)!,
+                "mayClientSendStateEvent",
+            ).mockReturnValue(false);
+
+            const { queryByLabelText } = getComponent({ mxEvent: alicesMessageEvent });
+            expect(queryByLabelText("Pin")).toBeFalsy();
+        });
+
+        it("should renders pin button", () => {
```
Comment (dbkr) [inline test/components/views/messages/MessageActionBar-test.tsx:498]:
```suggestion
        it("should render pin button", () => {
```

File: test/components/views/messages/MessageActionBar-test.tsx [hunk 2]
```diff
@@ -468,4 +471,33 @@ describe("<MessageActionBar />", () => {
...
+        });
+
+        afterEach(() => {
+            jest.spyOn(
+                room.getLiveTimeline().getState(EventTimeline.FORWARDS)!,
+                "mayClientSendStateEvent",
+            ).mockRestore();
+        });
+
+        it("should not renders pin button when user can send state event", () => {
```
Comment (dbkr) [inline test/components/views/messages/MessageActionBar-test.tsx:488]:
```suggestion
        it("should not render pin button when user can't send state event", () => {
```

I this this was the wrong way around?

File: test/utils/PinningUtils-test.ts [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+import { EventTimeline, EventType, IEvent, MatrixClient, MatrixEvent, Room } from "matrix-js-sdk/src/matrix";
+import { mocked } from "jest-mock";
+
+import { createTestClient } from "../test-utils";
+import PinningUtils from "../../src/utils/PinningUtils";
+import SettingsStore from "../../src/settings/SettingsStore";
+import { canPinEvent, isContentActionable } from "../../src/utils/EventUtils";
+import { ReadPinsEventId } from "../../src/components/views/right_panel/types";
+
+// The mock factory returns a mocked function
```
Comment (dbkr) [inline test/utils/PinningUtils-test.ts:26]: I'm not sure what this comment is talking about?

**Review by dbkr — APPROVED** [review]
Generally looks good. Good, thorough set of tests.

File: test/utils/PinningUtils-test.ts (hunk 3 above)
Comment (florianduros) [reply test/utils/PinningUtils-test.ts:26]: It should be deleted :)

**Pull request merged**